# **Divine**
**Di**sease-causing genes/**v**ariant prioritization in cl**in**ical whole **e**xome sequencing data

Divine is designed to make molecular diagnosis with high-throughput whole exome sequencing data more effective. Using both patient phenotypic information and genetic variants, Divine applies a machine learning technique to systematic multi-omic data integration such that it can prioritize a gene list causing a patient disease.

## Website

https://github.com/cjhong/divine

# Input
- a text file containing Human Phenotype Ontology (HPO) IDs (e.g., HP:0002307) that describe patient's clinical features

- or (and) VCF file

# Algorithm
- Refer to "Divine: Disease-causing genes/Variant prioritization in clINical whole Exome sequencing data", Hong. *et. al.*

# Output
- When only HPO IDs are given, Divine generates a prioritized gene list and an inferred disease list.

- If VCF file (or with HPO IDs), Divine also generates an annotated variant table with a ranking score in Microsoft Excel format.

# Developer note
- This is the README file for the Divine program. Divine is in active development. Please check the above website for newer available versions. Please contact the developers at changjin.hong@gmail.com to report any problems or for additional help.

# Setup

## Prerequisite:
- Linux, 4 GB RAM, and 130 GB available hard disk space
- python 2.7+
- make sure that 'pip', 'wget' (with Internet connection), 'grep', 'awk', and 'sort' command in $PATH
 
## Python modules to be installed
Divine requires the following modules but, during the setup process, the modules will be installed automatically if necessary.

- *fastsemsim-0.9.4: https://sites.google.com/site/fastsemsim
- *hgvs: https://github.com/counsyl/hgvs
- *hpo_similarity: https://github.com/jeremymcrae/hpo_similarity
- ConfigParser, backports-abc, backports.ssl-match-hostname, certifi, decorator, matplotlib, networkx, nose, numpy, pandas, pygr, pyparsing, pysam, python-dateutil, pytz, scikit-learn, singledispatch, six, tornado, xlwt

- [*]: the python modules are already included in the Divine package.

## Install

download divine source codes from github
```
$ git clone https://github.com/cjhong/divine.git
```

### Option 1: fresh install or to upgrade Divine resource package

It requires downloading 15GB files and so be patient!
```
$ setup.py --install --update_db
```

### Optional 2: only reinstall python modules of dependency
```
$ setup.py --install
```

## Configuration
- Get environment variables (DIVINE,GCN,GCN_DATA_DIR,GCN_DB_DIR,GCN_LOGFILE,PATH,PYTHONPATH) at the end of the installation setup message. Then,

- Add the environment variables into your shell configuration (e.g., $HOME/.bash_profile, $HOME/.profile, or $HOME/.cshrc)

- Make sure that the new configuration is applied by
  - logoff and login your account

  - running
  ```
  $ source $HOME/your_shell_config_file
  ```

## Uninstall
### Optional 1: only uninstall python modules of dependency
```
$ setup.py --uninstall
```

### Optional 2: also, uninstall resource files
```
$ setup.py --uninstall --remove_db
```

# Usage
## Input: a text file containing HPO IDs

First, visit either http://compbio.charite.de/phenomizer or https://mseqdr.org/search_phenotype.php. Enter the patient phenotype terms, description, or specific keywords you think important. Get the best matching HPO IDs. Paste the HPO IDs in the format of HP:XXXXXXX (e.g., HP:0002307) into a text file line by line and save it as a text file (e.g., P0001.hpo)

For example, an HPO file looks like
```
$ cat P0001.hpo
#my_patient_ID
HP:0002307
HP:0000639
HP:0001252
HP:0100543
HP:0002120
HP:0000708
HP:0001344
HP:0008872
HP:0000510
HP:0001513
HP:0006979
HP:0000752
HP:0012469
HP:0000577
HP:0001010
HP:0006887
HP:0002650
HP:0005469
HP:0002312
HP:0010808
HP:0002136
HP:0200085
HP:0002311
```

## Input: a VCF file
Any VCF file following the standard format (e.g., https://samtools.github.io/hts-specs/VCFv4.2.pdf). It can be generated by GATK, samtools, freebayes, or anything else.

## Output

### Known disease matching by patient phenotypes
`hpo_to_diseases.tsv`: From an input HPO file, Divine prioritize which disease the patient likely has. The output format is
```
$ cat hpo_to_disease.tsv
#disease_ID	genes	score[FunSimMax]
OMIM:101600	FGFR1,FGFR2	0.000911782
OMIM:101200	FGFR2	0.000674322
:
:
```
### Ranked gene by only phenotypic information

`rank_pheno_gene.tsv`: One gene is associated with more than two diseases. This output reformat `hpo_to_diseases.tsv` by sorting the phenotype matching score by gene. The output format is

```
$ cat rank_pheno_gene.tsv
#gene   phenotypic_score
FGFR1   0.000860439
FGFR2   0.000860439
FGFR3   0.000628328
TWIST1  0.000605784
:
:
```

### Annotated VCF files
- `divine.vcf`

	This is a VCF file annotated by Varant where we improve the original method significantly. Refer to Varant website (http://compbio.berkeley.edu/proj/varant).

- `vfilter.vcf`

	This is a subset of the VCF file above (i.e., `divine.vcf`). This VCF file contains variants after filtering out all variants that is either located in an intergenic region or its MAF is high or in repeated region.

### Prioritized genes and Microsoft excel file

- `rank.tsv`

	Ranking score assigned to each gene appeared into `vfilter.vcf`

- `divine.xls`

	This is an excel table by generated by Varant. Divine annotates a ranking score per gene so that you can sort the variant list by the ranking score (refer to 'Comments_on_Genes', an annotated column in the 2nd tab, VCF).

### Log files

- check log files under a subdirectory `logs` in an output directory

## Cases

- When HPO file is only given,

```
divine.py -q dir_to_the_hpo/P0001.hpo -o dir_to_output/P0001
```

- When VCF file is only given,

```
divine.py -v dir_to_the_vcf/P0002.vcf -o dir_to_output/P0002
```

- When both HPO file and VCF file are given,

```
divine.py -q dir_to_the_hpo/P0003.hpo -v dir_to_the_vcf/P0003.vcf -o dir_to_output/P0003
```

## Examples
```
cd $DIVINE/gcn/bin/prioritize/examples
./runme_pfeisffer.sh
./runme_millerSyndrome.sh
./runme_angelman.sh
```

# Help
```
usage: divine.py [-h] [-q HPO_QUERY_FN] [-v VCF] [-o OUT_DIR]
                 [-c VCF_FILTER_CFG] [-d EXP_TAG] [-i INDEL_MODE]
                 [-r SEED_RATE] [-e REF_EXON_ONLY] [-C CADD] [-H HGMD]
                 [-k VKNOWN] [-t CAPKIT] [--reuse]

Divine (v0.1.1) [author:changjin.hong@gmail.com]

optional arguments:
  -h, --help            show this help message and exit
  -q HPO_QUERY_FN, --hpo HPO_QUERY_FN
                        Input patient HPO file. A file contains HPO IDs (e.g.,
                        HP:0002307), one entry per line. Refer to
                        http://compbio.charite.de/phenomizer or
                        https://mseqdr.org/search_phenotype.php
  -v VCF, --vcf VCF     input vcf file
  -o OUT_DIR, --out_dir OUT_DIR
                        output directory without white space. If not exist,
                        the directory will be created.
  -c VCF_FILTER_CFG, --vcf_filter_cfg VCF_FILTER_CFG
                        vcf filter configuration file [None]
  -d EXP_TAG, --exp_tag EXP_TAG
                        specify experiment tag without white space. The tag
                        will be contained in the output file name.[None]
  -i INDEL_MODE, --indel INDEL_MODE
                        the level of fidelity of indell call in VCF, [1]:low
                        (e.g., samtools), 2:high (GATK haplotype caller)
  -r SEED_RATE, --seed_rate SEED_RATE
                        the rate of choosing matched diseases from top for
                        gene enrichment [0.0015]; set to 0. to disable
  -e REF_EXON_ONLY, --ref_exon_only REF_EXON_ONLY
                        the annotation process only runs on RefSeq coding
                        regions 0:No, [1]:Yes
  -C CADD, --cadd CADD  use CADD prediction score, 0:No, [1]:Yes
  -H HGMD, --hgmd HGMD  enable HGMD (requires a license), [0]:No, 1:Yes
  -k VKNOWN, --vknown VKNOWN
                        apply variant-level pathogenic annotation (e.g.,
                        either ClinVar or HGMD) to prioritization strategy,
                        0:No, [1]:Yes
  -t CAPKIT             capture kit symbol [SureSelect_V6]
  --reuse               Reuse previous annotation file (divine.vcf) if it is
                        available [False]
```

# FAQ

- Q.1: I have a VCF file generated from either WGS or WES dataset and the number of variants is a lot! It seems that Divine is slow and I want to have a result as soon as possible.

- A.1: Depending on your hardware specification and the number of variants in an input VCF file, the computational time varies. In my computer setting (e.g., Intel Core 2 Duo @ 2.93GHz), it takes 30 min to handle 37,000 variants. The computational bottleneck is directly associated with hard disk drive I/O. HDD connected to USB 3.0 or SDD (solid-state drive) can be helpful. Also, we are actively working on Divine to speed up the process. For now, Divine is optimized to analyze WES or targeted panel samples but not WGS. As a default option, Divine focuses on NCBI RefGene exon regions with +/-20bp flanking. However, be aware that it may not detect a pathogenic variant in intergenic region or up/downstream. For a VCF file containing a small number of variants, we suggest `-e 0` instead. 

- Q.2: I don't like a default filtering scheme used in Divine. I want my own filtering strategy (e.g., include a certain flag in FILTER; not to use ExAC; 0.03 for MAF cutoff).

- A.2: open $DIVINE/gcn/config/filterconf.txt and edit the configuration file. For example, set `excl=LowQual` to filter out all LowQual in the VCF file. Later, we will provide more comprehensive instruction for this file. The default configuration is

```
[fltr]
excl=LowQual
[infoflag]
excl=DB137
[infoval]
kgaf=yes
espaf=yes
exacaf=yes
splice_dist=20
hgmd_filter=2
regulome=no
[reg]
incl=CodingExonic:NonCodingExonic:CodingIntronic:NonCodingIntronic
[freq]
incl=0.01
[freq_cli]
incl=0.05
[gid]
min=0.1
```

- Q.3: I purchase HGMD professional license and how can I use this feature?

- A.3: Contact to changjin.hong@gmail.com

- Q.4: Previously, I ran Divine on a patient sample dataset which took so long. Now, I want to prioritize genes with a different setting (e.g., filtering condition or a different option in `divine.py`) and I am sure that I didn't update Divine/GCN database. How I can make the repeat analysis faster?

- A.4: Unless Divine database is changed, or purchase HGMD license newly, or divine.vcf created previously is corrupted, you can reuse the previous annotated VCF file using `--reuse`. If you want to compare a new result with the previous one, try

```
$ $DIVINE/gcn/bin/prioritize/divine.py -q dir_to_the_hpo/P0005.hpo \
   -v dir_to_the_vcf/P0005.vcf --reuse -d rev -o dir_to_the_output/P0005
```

- Q.5: For the same input and configuration, will Divine generate the same output?

- A.5: Yes, Divine is deterministic. There is no randomness in the analysis. Also, Divine generates all log files so that you can audit/trace the previous experiment results and database maintenance any time.

- Q.6: I think database or annotation Divine provides is outdated. I would like to keep all the latest database so that I can improve my diagnosis result at best.

- A.6: Divine uses Varant as an annotation framework. We will frequently update the database. Each resource package will be archived. Furthermore, we plan to provide a stream pipeline to update annotation database.

- Q.7: I know Divine is designed for germline or constitutional disease samples. Can I use Divine for somatic mutation (cancer sample) analysis?

- A.7: We will work on this feature as well.

- Q.8: I have a patient dataset but I am concerning if Divine collects the patient dataset without my permission and send it somewhere.

- A.8: This standalone package never collect any user information or input dataset the user works. But, we need your feedback and bug report!

- Q.9: Does Divine support an analysis for trio samples (e.g., proband, mother, and father)?

- A.9: Yes, if VCF file contains multiple samples, then trio analysis can be done in `divine.xls`. The next release of Divine will provide more systematic trio analysis.

- Q.10: I don't have Linux computer. Can Divine run on MacOS or Windows?

- A.10: Not now but we are currently working on the compatibility issue.

- Q.11: What is the memory requirement?

- A.11: In Varant annotation step, memory usage peaks around 1.5GB.

- Q.12: My company wants to use Divine.

- A.12: Divine is free for academic or research purpose. Contact to changjin.hong@gmail.com for a commercial use.

- Q.13: I follow this README file and help. My Divine run fails. Can you help me?

- A.13: Three sample scripts with datasets in `$DIVINE/gcn/bin/prioritize/examples` are available. Use the examples as a template. All three scripts should work if your installation was successful. If the problem is not resolved, send me a log file (e.g., `divine_err.log`) by running,

```
$ your_divine_command 2>&1 | tee divine_err.log
```

- Q.14: Can Divine detect a gene previously never known to be associated with a certain disease?

- A.14: This is challenging task but it is essential. Currently, Divine uses gene ontology enrichment and also it uses a system biology approach (heat diffusion model) over STRING protein functional network.

# Change Log
- v.0.1.1 (June 15 2016)
	- Original release, Clinvitae(http://clinvitae.invitae.com/) added

# License
GNU GENERAL PUBLIC LICENSE
https://www.gnu.org/licenses/gpl-3.0.en.html

# Disclaimer
Not intended for direct clinical application. Divine suggests an order of genes to be inspected so that it can make molecular diagnosis effective. The validation is the responsibility of the user. Neither Divine developer nor any software module integrated is responsible for clinical actions that may result from the use of this software. By using this tool, the user assumes all responsibility for any information that may be generated.

# Reference
- Varant: http://compbio.berkeley.edu/proj/varant/publication.html
- Divine: Disease-causing genes/Variant prioritization in clinical whole exome sequencing data, Changjin Hong, David Yang, and Tae-Hyun Hwang (in publication)

# Contact
- Changjin Hong, Ph.D (changjin.hong@gmail.com)
